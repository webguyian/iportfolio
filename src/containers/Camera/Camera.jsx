import React from 'react';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import Link from 'components/Link/Link';
import RoundedButton from 'components/Button/RoundedButton';
import Text from 'components/Text/Text';
import CameraFilters from './CameraFilters/CameraFilters';

import { useCamera, useCameraFilters } from 'modules/camera/hooks';

const Camera = () => {
  const [elements, actions, controls] = useCamera();
  const { flash, timer, filter, swap } = controls;
  const filterState = filter.state;
  const [filters, activeFilter, onFilterClick] = useCameraFilters(
    elements,
    filterState
  );
  const baseClass = 'camera-app';
  const videoClass = `${baseClass}-video`;
  const controlClass = `${baseClass}-control`;
  const getControlStateClass = control => `${controlClass}--${control.state}`;

  return (
    <div className={baseClass}>
      <div className={`${baseClass}-top-bar`}>
        <Button
          className={classNames(controlClass, getControlStateClass(flash))}
          icon="bolt"
          size="2x"
          disabled={flash.state === 'disabled'}
          onClick={flash.onClick}
        >
          Set flash
        </Button>
        <Button
          className={classNames(controlClass, getControlStateClass(timer))}
          icon="history"
          size="2x"
          disabled={timer.state === 'disabled'}
          onClick={timer.onClick}
        >
          Set timer
        </Button>
        <Button
          className={classNames(controlClass, getControlStateClass(filter))}
          icon="swatchbook"
          size="2x"
          disabled={filterState === 'disabled'}
          onClick={filter.onClick}
        >
          Set filter
        </Button>
      </div>
      <div className={`${videoClass}-container`}>
        <canvas
          className={`${baseClass}-canvas`}
          ref={elements.canvas}
          width="375"
          height="425"
        ></canvas>
        <canvas
          className={`${baseClass}-canvas-overlay`}
          ref={elements.overlay}
          width="375"
          height="425"
        ></canvas>
        <video
          className={videoClass}
          ref={elements.video}
          autoPlay
          playsInline
          onPlay={actions.onPlay}
        ></video>
      </div>
      {filterState === 'active' ? (
        <CameraFilters
          filters={filters}
          activeFilter={activeFilter}
          onFilterClick={onFilterClick}
        />
      ) : (
        <Text className={`${baseClass}-label`} element="p" type="accessible">
          Grant camera access to take photos
        </Text>
      )}
      <div className={`${baseClass}-bottom-bar`}>
        <Link to="/photos">
          {elements.image ? (
            <img className={`${baseClass}-image`} src={elements.image} />
          ) : (
            <div className={`${baseClass}-image-placeholder`} />
          )}
        </Link>
        <RoundedButton
          disabled={elements.button === 'disabled'}
          onClick={actions.onClick}
        >
          <Text type="accessible">Capture snapshot</Text>
        </RoundedButton>
        <Button
          icon="sync"
          size="3x"
          disabled={swap.state === 'disabled'}
          onClick={swap.onClick}
        >
          Reverse camera
        </Button>
      </div>
    </div>
  );
};

export default Camera;
