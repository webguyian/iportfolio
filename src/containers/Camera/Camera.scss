/* stylelint-disable a11y/no-outline-none */
.camera-app {
  @include stretch-column-nowrap;

  &-top-bar,
  &-bottom-bar {
    display: flex;
    align-items: center;
    justify-content: space-around;
    padding: $gutter 0;

    .ui-icon {
      transition: all 0.4s;
      color: $white;
    }
  }

  &-control {
    &--active {
      > .ui-icon {
        color: $orange;
      }
    }
  }

  &-video-container {
    display: flex;
    position: relative;
    flex-flow: column wrap;
    margin-bottom: $gutter;
  }

  &-canvas-overlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }

  &-video {
    display: none;
  }

  &-label {
    @include type-display;
    margin: $gutter;
    text-align: center;
  }

  &-bottom-bar {
    margin: auto 0 0;
  }

  &-image,
  &-image-placeholder {
    width: 70px;
    height: 70px;
    background-color: $gray-dark;
  }

  .ui-btn--rounded {
    @include circle(80px);
    @include inner-circle(60px, $white, $gray-dark);
    margin: 0;
    transition: background-color 0.4s;
    background-color: $white;

    &:hover,
    &:focus {
      background-color: $gray-light;

      &::before {
        background-color: $gray-light;
      }
    }

    &:focus {
      outline: 0;
    }

    &[disabled] {
      background-color: $gray-medium;

      &::before {
        background-color: $gray-medium;
      }
    }
  }

  .ui-btn--with-icon[disabled] {
    > .ui-icon {
      color: $gray-medium;
    }
  }
}
