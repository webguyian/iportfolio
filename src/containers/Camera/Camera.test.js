import React from 'react';
import { render } from '@testing-library/react';

import Camera from './Camera';

describe('<Camera />', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Camera />);

    expect(asFragment()).toMatchSnapshot();
  });
});
