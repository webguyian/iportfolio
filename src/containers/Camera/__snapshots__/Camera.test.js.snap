// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Camera /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="camera-app"
  >
    <div
      class="camera-app-top-bar"
    >
      <button
        class="ui-btn ui-btn--with-icon camera-app-control camera-app-control--disabled"
        disabled=""
        type="button"
      >
        <i
          class="ui-icon ui-icon-bolt fas fa-bolt fa-2x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Set flash
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon camera-app-control camera-app-control--enabled"
        type="button"
      >
        <i
          class="ui-icon ui-icon-history fas fa-history fa-2x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Set timer
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon camera-app-control camera-app-control--enabled"
        type="button"
      >
        <i
          class="ui-icon ui-icon-swatchbook fas fa-swatchbook fa-2x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Set filter
        </span>
      </button>
    </div>
    <div
      class="camera-app-video-container"
    >
      <canvas
        class="camera-app-canvas"
        height="425"
        width="375"
      />
      <canvas
        class="camera-app-canvas-overlay"
        height="425"
        width="375"
      />
      <video
        autoplay=""
        class="camera-app-video"
        playsinline=""
      />
    </div>
    <p
      class="ui-text ui-text--accessible camera-app-label"
    >
      Grant camera access to take photos
    </p>
    <div
      class="camera-app-bottom-bar"
    >
      <mock-link
        to="/photos"
      >
        <div
          class="camera-app-image-placeholder"
        />
      </mock-link>
      <button
        class="ui-btn ui-btn--rounded"
        type="button"
      >
        <span
          class="ui-text ui-text--accessible"
        >
          Capture snapshot
        </span>
      </button>
      <button
        class="ui-btn ui-btn--with-icon"
        disabled=""
        type="button"
      >
        <i
          class="ui-icon ui-icon-sync fas fa-sync fa-3x"
        />
        <span
          class="ui-text ui-text--accessible"
        >
          Reverse camera
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
